### ISSUE TITLE
question: suppress module path expansion

### ISSUE THREAD

**Issue by nilsbecker** [the-lambda-church/merlin#447]
Merlin often expands paths to the longest possible form, for instance, `Containers.CCVector` instead of `Vector` and  `Containers.CCList` instead of `List` when `Containers` is opened at the top level. I really do not want these superfluous long paths in my source. Is there a way to prevent this?

**Comment by def-lkb** [comment]
When printing types? You can activate `-short-paths` by passing this flag or adding it in `.merlin`:
```
FLG -short-paths
```

**Comment by nilsbecker** [comment]
actually, when using tab-expansion while editing (i use supertab for that). i'll try `-short-paths`

**Comment by def-lkb** [comment]
So you would like Containers.CCV<tab> to be rewritten to CCVector automatically when Containers is opened?
That can be done the API will need some rework. But why not start with CCV<tab> directly?

This is related to #396 for which I have a prototype in the work.

**Comment by nilsbecker** [comment]
well actually, the shortest way to use `Containers.CCVector` after opening `Containers` is simply as `Vector`. note that `Containers` also defines extensions to many standard library modules, which are included and then shadowed. so after opening `Containers`, the `List` module contains some functions that actually come from the std lib and some extra ones which come from `Containers.CCList` but all are available as `List.soandso`; and this is how the library is meant to be used. i think with batteries it's the same.

so basically, i would want to always use the shortest version when expanding a name. actually, #396 looks really useful too!

**Comment by def-lkb** [comment]
But then, just don't type `Containers.`. During completion, merlin will show you the short names too.
I think I am missing your point.

**Comment by nilsbecker** [comment]
ok, i think i wasn't clear. what i do is i type
`let aa = List.app<tab>` and this is completed as 
`let aa = Containers.List.append` which is the same as `List.append` since `Containers` was opened. i really wanted `List.append`.

**Comment by def-lkb** [comment]
Oh I see, something wrong is going on.
Merlin should keep the short form.

I will try to reproduce, can you share more information about your setup?

**Comment by nilsbecker** [comment]
ok, here is the merlin section of my .vimrc


    " merlin runtime path settings
    let g:opamshare = substitute(system('opam config var share'),'\n$','','''')
    execute "set rtp+=" . g:opamshare . "/merlin/vim"
    " redundant but what the hell.
    execute "helptags " . g:opamshare . "/merlin/vim/doc"
    
    " merlin extras
    let g:syntastic_ocaml_checkers=['merlin']
    let g:syntastic_auto_loc_list=2
    let g:merlin_completion_arg_type = "always"
    augroup ocamlmerlin
    	au!
    	au FileType ocaml call SuperTabSetDefaultCompletionType("<c-x><c-o>")
    	" list the expected argument type when completing
    	" highlighting of type selections; only works for light bg...
    	" problems in the terminal
    	"au FileType ocaml hi EnclosingExpr ctermbg=17 guibg=Green
    				"\ ctermfg=bg guifg=bg
    	" useful bindings
    	au FileType ocaml nmap <buffer> <localleader>T :MerlinGrowEnclosing<cr>
    	au FileType ocaml nmap <buffer> <localleader>yt :MerlinYankLatestType<cr>
    	au FileType ocaml nmap <buffer> <localleader>l :MerlinLocate<cr>
    	au FileType ocaml nmap <buffer> <localleader>r <plug>(MerlinRename)
    	au FileType ocaml nmap <buffer> <localleader>o :MerlinOccurrences<cr>
    	au FileType ocaml nmap <buffer> <localleader>d :MerlinDestruct<cr>
    	au FileType ocaml nmap <buffer> <localleader>e <plug>(MerlinErrorCheck)
    	au FileType ocaml nmap <buffer> <localleader>ce :lclose<cr>
    	" set the local ocamlbuild directories automatically
    	"au FileType ocaml silent SourcePath %:p:h
    	"au FileType ocaml silent BuildPath %:p:h/_build
    	" the text object for navigating types
    	au FileType ocaml let g:merlin_textobject_shrink='m'
    	au FileType ocaml let g:merlin_textobject_grow='M'
    	au CursorMoved *.ml MerlinClearEnclosing
    	au CursorMoved *.mli MerlinClearEnclosing
      au FileType ocaml nmap <localleader>*  <Plug>(MerlinSearchOccurrencesForward)
      au FileType ocaml nmap <localleader>#  <Plug>(MerlinSearchOccurrencesBackward)
    augroup END

do you need anything else?

**Comment by nilsbecker** [comment]
another bit of information: if `IH` is generated via the functor `CCHashtbl.Make` then if i type
`IH.rep<tab>` i get `CCHashtbl.Make.replace` which is also not helpful...

**Comment by def-lkb** [comment]
I have no idea how this is possible. Which version of merlin are you using?

**Comment by nilsbecker** [comment]
`:MerlinVersion` gives Merlin toolkit version 2.2 for Ocaml 4.02.3. i installed merlin via opam.

**Comment by nilsbecker** [comment]
my `.merlin`:

    PKG batteries
    PKG containers
    PKG sequence
    PKG gsl
    S .
    B _build
    FLG -short-paths

**Comment by nilsbecker** [comment]
i could post a gist with the module and .mli if that would help...

**Comment by def-lkb** [comment]
Don't worry, I'll try my best and ask if I need more :)

**Comment by def-lkb** [comment]
I know what you can do :):
Run merlin after `export MERLIN_LOG=merlin.log`.
It will log communication between vim and merlin.
Trigger the problem once (to not make the log too big), close and share merlin.log.

(Be aware that your source code will be partially included, send by mail or filter "tell" lines if you prefer, e.g. `grep -v '"tell"' merlin.log`).

**Comment by nilsbecker** [comment]
ok, here are the lines with 'tell'. i started vim and invoked the completion once. 

{"time":2,"level":"info","section":"general","title":"monitor start","content":["general","info"]}
{"time":2,"level":"info","section":"general","title":"available logging sections","content":["general","destruct","dot_merlin","typedtrie","project_load","parser","locate","recover","protocol","explain","main"]}
{"time":2,"level":"info","section":"general","title":"monitor start","content":["protocol","info"]}
{"time":2,"level":"info","section":"protocol","title":"input","content":{"delta":0,"body":["reset","auto","/Users/nbecker/Tresors/Code/bd/bd.ml"]}}
{"time":5,"level":"info","section":"protocol","title":"output","content":{"delta":2,"body":["return",{"cursor":{"line":1,"col":0},"marker":false}]}}
{"time":5,"level":"info","section":"protocol","title":"input","content":{"delta":0,"body":["project","get"]}}
{"time":5,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"result":["/Users/nbecker/Tresors/Code/bd/.merlin"]}]}}
{"time":5,"level":"info","section":"protocol","title":"input","content":{"delta":0,"body":["reset","auto","/Users/nbecker/Tresors/Code/bd/bd.ml"]}}
{"time":5,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1,"col":0},"marker":false}]}}
{"time":5,"level":"info","section":"protocol","title":"input","content":{"delta":0,"body":["seek","exact",{"line":1,"col":0}]}}
{"time":5,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1,"col":0},"marker":false}]}}
{"time":5,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1,"col":0},"marker":false}]}}
{"time":21,"level":"info","section":"protocol","title":"output","content":{"delta":15,"body":["return",{"cursor":{"line":1006,"col":0},"marker":false}]}}
{"time":21,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1006,"col":0},"marker":true}]}}
{"time":21,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1006,"col":0},"marker":false}]}}
{"time":225,"level":"info","section":"protocol","title":"input","content":{"delta":203,"body":["project","get"]}}
{"time":225,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"result":["/Users/nbecker/Tresors/Code/bd/.merlin"]}]}}
{"time":225,"level":"info","section":"protocol","title":"input","content":{"delta":0,"body":["seek","exact",{"line":1005,"col":0}]}}
{"time":225,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1003,"col":5},"marker":false}]}}
{"time":225,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1001,"col":9},"marker":false}]}}
{"time":225,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1006,"col":0},"marker":false}]}}
{"time":225,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1006,"col":0},"marker":true}]}}
{"time":225,"level":"info","section":"protocol","title":"output","content":{"delta":0,"body":["return",{"cursor":{"line":1006,"col":0},"marker":false}]}}

**Comment by def-lkb** [comment]
Ok, it seems merlin is never used for completion.

**Comment by def-lkb** [comment]
So, the problem comes from somewhere else, which is a bit reassuring.

**Comment by nilsbecker** [comment]
oh! this means i mis-configured the completion somehow? do you know how to fix it?

**Comment by def-lkb** [comment]
Check the value of `:set omnifunc?`.
If you try C-x C-o directly, do you have a better behavior?

**Comment by nilsbecker** [comment]
aha! i think we have the culprit. it's `ocpindex#complete`. sorry for making a fuss. it seems that this line in my .vimrc:

    execute "au FileType ocaml set rtp+=" . g:opamshare . "/ocp-index/vim"

sets the omnifunc!? is ocp-index redundant or conflicting with merlin in general?

**Comment by trefis** [comment]
Yes to both.

2015-10-13 16:41 GMT+01:00 nilsbecker <notifications@github.com>:

**Comment by nilsbecker** [comment]
ok, removed the line, completion works as expected again! thanks for the help! i guess this can be closed then...
